### PR TITLE
RAG-2nd2end-revamp

### PR DESCRIPTION
**Pull request by shamanez**
same as [shamanez:rag-retriever-end2end](https://github.com/huggingface/transformers/pull/11655) PR. In the previous version, I got some version control problems.

Really sorry to come up with duplicate PRs :(.   

@lhoestq  @patrickvonplaten 

I conducted an experiment to check the difference and added a simple test run code. 

Finally I added two test functions to test_modeling_rag.py and test_retrieval_rag.py.  

image

### CODE REVIEW

File: examples/research_projects/rag-end2end-retriever/README.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# End-to-End finetuning of RAG (including DPR retriever) for Question Answering.
+
+This finetuning script is actively maintained by [Shamane Siri](https://github.com/shamanez). Feel free to ask questions on the [Forum](https://discuss.huggingface.co/) or post an issue on [GitHub](https://github.com/huggingface/transformers/issues/new/choose) and tag @shamanez.
```
Comment (patrickvonplaten) [inline examples/research_projects/rag-end2end-retriever/README.md:3]: Thanks a lot for adding this, that's great!

**Review by patrickvonplaten — APPROVED** [review]
This looks good to me - thanks for adding some tests! 
@lhoestq can you take a quick look?

**Review by lhoestq — APPROVED** [review]
Amazing ! Good job @shamanez :)

I took a look mainly at the changes in transformers/models/rag and the indexing in the research_project/rag-end2end-retriever folder and it's all good to me !

I just noticed two typos and one minor change to do for the return_tensors parameter of the retriever

**Comment by shamanez** [comment]
I added all the minor changes and I would like to thank @patrickvonplaten and @lhoestq for the enormous amounts of support and advice.  :)

**Comment by sb1992** [comment]
Hey 
Thanks for giving this end to end version. Am trying to run it and see its performance on my own dataset (much smaller and domain specific to see if get performance gains with end to end) but at the moment it is throwing a Pickling error with dummy dataset in the code. Am still stuck trying to understand how to fix this. Any idea how this can be dealt with

  
      File "/home/shraeyb/anaconda3/envs/new_st/lib/python3.7/site-packages/ray/cloudpickle/cloudpickle_fast.py", line 580, in dump
        return Pickler.dump(self, obj)
      File "pyarrow/io.pxi", line 1021, in pyarrow.lib.Buffer.__reduce_ex__
    AttributeError: module 'pickle' has no attribute 'PickleBuffer

Cheers

**Comment by shamanez** [comment]
Could you please add this as a new issue. Also I would like to see the
entire error log. Seems like something is wrong with your RAY installation.

**Comment by sb1992** [comment]
Sure, just created a new issue with full log

**Comment by g-karthik** [comment]
@shamanez looks like the Google Drive link you provided in the `README.md` is dead now, I get an error. Can you please take a look?

**Pull request merged**